glib/src/object.rs

```diff
@@ -1756,12 +1820,25 @@ impl<T: ObjectType> ObjectExt for T {
     {
         let signal_name: &str = signal_name.into();
         let type_ = self.get_type();
-
-        let (signal_id, signal_detail) = SignalId::parse_name(signal_name, type_, true)
+        let (signal_id, details) = SignalId::parse_name(signal_name, type_, true)
             .ok_or_else(|| bool_error!("Signal '{}' of type '{}' not found", signal_name, type_))?;
-        let signal_query = signal_id.query();
+        self.connect_unsafe_id(signal_id, Some(details), after, callback)
+    }
 
+    unsafe fn connect_unsafe_id<F>(
+        &self,
+        signal_id: SignalId,
+        details: Option<Quark>,
+        after: bool,
+        callback: F,
+    ) -> Result<SignalHandlerId, BoolError>
+    where
+        F: Fn(&[Value]) -> Option<Value>,
+    {
+        let signal_query = signal_id.query();
+        let type_ = self.get_type();
```

Comment:
You could check here that the query gives a type that is compatible with `self`

Looks good to me otherwise, thanks!